### PR TITLE
[fix][train][step-wise] Broadcast step-wise advantage with each step's own response_mask

### LINKED ISSUE #1492: [train] Incorrect advantage assignment for step_wise_trajectores in RayPPOTrainer?
I've been having some stability issues when training with step_wise_trajectories. I see in the RayPPOTrainer we are doing 

```
traj_ids = (
    torch.cat([torch.tensor([False], device=is_last_step.device), is_last_step[:-1]]).int().cumsum(dim=0)
)
num_groups = traj_ids[-1].item() + 1
assert num_groups == len(
    last_step_advantages
), f"number of groups {num_groups} doesn't match the number of trajectories as given by `is_last_step` {len(last_step_advantages)}. The `is_last_step` tensor is likely malformed"
advantages = last_step_advantages[traj_ids]
returns = last_step_returns[traj_ids]
```

[link](https://github.com/NovaSky-AI/SkyRL/blob/main/skyrl/train/trainer.py#L823)

But assigning `advantages = last_step_advantages[traj_ids]` doesn't seem quite right to me since `last_step_advantages` is a `scalar * response_mask` of the last step. This means the previous steps in a trajectory are getting an advantage with a different response mask. This is especially problematic if the last step `response_ids` is short and the prior steps see much longer obs_ids (thus there is virtually no overlap in action tokens due to the left padding). Shouldn't we be getting the advantage of the last step as a scalar and multiplying that by `data["response_mask"]`? 

When making this fix I'm seeing much better stability, but do let me know if I'm looking at this correctly. Also happy to open a small PR if that helps.

### PR DESCRIPTION
**Pull request by CharlieFRuan**
In step-wise training, `advantages = last_step_advantages[traj_ids]` copied the last step's already-masked tensor verbatim to every step of the trajectory. Because each step has its own right-aligned response_mask at its own token positions (prompts and responses vary in length across turns), the scalar advantage landed at the last step's token positions — which almost never overlap with the earlier step's own response positions. Earlier steps silently received advantages = 0 at their loss_mask=1 positions, so roughly (N-1)/N of step-samples are trained incorrectly.

Fix:
  - Call the estimator with an all-ones mask so it returns the scalar advantage at every position (outcome-based estimators produce `scalar * mask`).
  - Broadcast by indexing into `traj_ids`, then multiply by each step's own `response_mask` so values land at the right token positions.
  - Reject `step_wise_trajectories=True` + `gae`/`reinforce++` in `validate_cfg`: temporal estimators produce per-token advantages that the scalar-broadcast model can't represent, and the step-wise path only sees the last step's slice anyway (no cross-step credit).
  - Add regression test `test_calc_advantages_and_returns_step_wise_broadcast` with a realistic right-aligned batch, plus a parametrized `test_validate_cfg_step_wise_estimator_compatibility` covering all 5 built-in estimators.
  - Document the outcome-only constraint in the step-wise tutorial.

Fixes https://github.com/NovaSky-AI/SkyRL/issues/1492.

### Testing

https://wandb.ai/sky-posttraining-uc-berkeley/skyrl-search-padding/reports/PR-1285-search-r1-PAD-left-only-baseline-after-PR-and-step-wise---VmlldzoxNjE3MTA0OQ

Since we will run into https://github.com/NovaSky-AI/SkyRL/issues/1501 at HEAD, we make this hacky hot fix, mainly the `pad_batch()` change: https://github.com/CharlieFRuan/SkyRL/commit/f2783a48006b9b4207345202388818568ba1c0df

```py
    if self.cfg.generator.step_wise_trajectories:
        n_samples = self.cfg.generator.n_samples_per_prompt
        pad_target = max(
            self.cfg.trainer.policy_mini_batch_size * n_samples,
            self.cfg.trainer.critic_mini_batch_size * n_samples,
            dp_size,
        )
    else:
        pad_target = dp_size
    pad_size = math.ceil(training_input.batch_size / pad_target) * pad_target - training_input.batch_size
```

We compare against the red curve from https://github.com/NovaSky-AI/SkyRL/pull/1285

Note that the actual curve should converge slower than this after we merge in https://github.com/NovaSky-AI/SkyRL/pull/1483, in which each mini batch will truly be one optimizer step. Currently we can have multiple optimizer step due to the number of turns can inflate the number of sequences in each mini batch.

image

The reward curve being higher is likely due to:

1) The fix we apply in this PR
2) In the previous PR (red line), we silently strip the amount of data and only train `num_mini_batches = len(data) // mini_batch_size`, which will no longer be an issue with the `pad_target` hacky fix above, or properly fixed after https://github.com/NovaSky-AI/SkyRL/pull/1483

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1507" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Review by SumanthRH — APPROVED** [review]

**Comment by SumanthRH** [comment]
Thanks! Would be good to get a E2E run!
